File: src/day20.py
import _ctypes


def get_object(obj_id):
    """ Inverse of id() function. But only works if the object is not garbage collected"""
    return _ctypes.PyObj_FromPtr(obj_id)


def day20_1(filepath):
    # Read the input file line by line
    with open(filepath, 'r') as f:
        input_list = [int(i) for i in f.readlines()]

    # Format: [original index, modified index, number value]
    original = input_list
    modified = [_ctypes.addressof(x) for x in input_list]  # Swap out values with the pointers to the original value!
    print(original)
    print(modified)
    list_len = len(original)

    for original_value in original:
        # Get the modified index of the original value
        modified_index = modified.index(id(original_value))
        modified_value = modified.pop(modified_index)
        # Then re-insert at the new location
        new_index = modified_index + original_value
        new_index = new_index % (list_len - 1)
        modified.insert(new_index, modified_value)

    result_list = [str(get_object(t)) for t in modified]
    print(" ".join(result_list))
    zero_index = result_list.index('0')
    answer = 0
    for x in [1000, 2000, 3000]:
        add = int(result_list[((zero_index + x) % list_len)])
        answer += add
    print(answer)

    # WTF
    buffer = [(idx, i) for idx, i in enumerate(input_list)]

    for idx, i in enumerate(input_list):
        old_idx = buffer.index((idx, i))

        buffer.remove((idx, i))
        buffer.insert((old_idx + i + len(input_list) - 1) %
                      (len(input_list) - 1), (-1, i))

    zero_idx = buffer.index((-1, 0))
    print(" ".join([str(f[1]) for f in buffer]))

    print(sum(buffer[(zero_idx + (i+1) * 1000) % len(buffer)][1]
          for i in range(3)))


if __name__ == '__main__':
    day20_1('data/test20_1.txt')
    # 4348 too low
    # day20_1('data/input20_1.txt')

File: src/day20.rs
use std::{fs, collections::HashSet};

pub fn day20_1(filepath: &str) -> Result<i32, std::io::Error> {
    // Read the input file line by line
    let input_str: String = fs::read_to_string(filepath)?;
    // Format: [original value, original index, modified index]
    let original: Vec<i32> = input_str
        .split("\n")
        .map(|(e)| e.parse::<i32>().unwrap())
        .collect();
    let mut modified: Vec<*const i32> = original.iter().map(|f| f as *const i32).collect();
    let list_len = original.len();

    // dbg!(&original);
    // dbg!(&modified);

    for original_value in original.iter() {
        // Get the modified index of the original value'
        // let test: Vec<_> = modified.iter().map(|p| {dbg!(*p); dbg!(original_value as *const i32)}).collect();
        let modified_index = modified.iter().position(|r| *r == original_value as *const i32).unwrap();
        // dbg!(modified_index);
        modified.remove(modified_index);

        // Then re-insert at the new location
        let mut new_index = modified_index as i32 + original_value;
        // dbg!(new_index);
        new_index = new_index.rem_euclid(list_len as i32 - 1);
        // dbg!(new_index);
        modified.insert(new_index as usize, original_value as *const i32);
        // dbg!(&modified);
    }

    // Print resulting list values
    // dbg!(&modified);
    let mut zero_index = 0;
    for (i, pointer) in modified.iter().enumerate() {
        unsafe {
            // dbg!(**pointer);
            if **pointer == 0 {
                zero_index = i as i32;
            }
        }
    }

    dbg!(&zero_index);

    let mut answer = 0;
    for x in vec![1000, 2000, 3000] {
        let pointer = modified[(zero_index + x).rem_euclid(list_len as i32) as usize];
        unsafe {
            // dbg!(*pointer);
            answer += *pointer;
        }
    }

    Ok(answer)
}

pub fn day20_2(filepath: &str) -> Result<i64, std::io::Error> {
    // Read the input file line by line
    let input_str: String = fs::read_to_string(filepath)?;
    // Format: [original value, original index, modified index]
    let original: Vec<i64> = input_str
        .split("\n")
        .map(|(e)| e.parse::<i64>().unwrap() * 811589153)
        .collect();
    let mut modified: Vec<*const i64> = original.iter().map(|f| f as *const i64).collect();
    let list_len = original.len();

    // dbg!(&original);
    // dbg!(&modified);
    for _ in 0..10 {

        for original_value in original.iter() {
            // Get the modified index of the original value'
            // let test: Vec<_> = modified.iter().map(|p| {dbg!(*p); dbg!(original_value as *const i64)}).collect();
            let modified_index = modified.iter().position(|r| *r == original_value as *const i64).unwrap();
            // dbg!(modified_index);
            modified.remove(modified_index);

            // Then re-insert at the new location
            let mut new_index = modified_index as i64 + original_value;
            // dbg!(new_index);
            new_index = new_index.rem_euclid(list_len as i64 - 1);
            // dbg!(new_index);
            modified.insert(new_index as usize, original_value as *const i64);
            // dbg!(&modified);
        }

    }

    // Print resulting list values
    // dbg!(&modified);
    let mut zero_index = 0;
    for (i, pointer) in modified.iter().enumerate() {
        unsafe {
            // dbg!(**pointer);
            if **pointer == 0 {
                zero_index = i as i64;
            }
        }
    }

    dbg!(&zero_index);

    let mut answer = 0;
    for x in vec![1000, 2000, 3000] {
        let pointer = modified[(zero_index + x).rem_euclid(list_len as i64) as usize];
        unsafe {
            // dbg!(*pointer);
            answer += *pointer;
        }
    }

    Ok(answer)
}


#[cfg(test)]
mod tests {
    // Note this useful idiom: importing names from outer (for mod tests) scope.
    use super::*;

    #[test]
    fn test_day20_1() {
        assert_eq!(day20_1("data/test20_1.txt").unwrap(), 3);
    }

    #[test]
    fn test_day20_2() {
        assert_eq!(day20_2("data/test20_1.txt").unwrap(), 1623178306);
    }
}